packages/compiler/test/checker/templates.test.ts

```diff
@@ -515,4 +515,230 @@ describe("compiler: templates", () => {
+        },
+        {
+          code: "invalid-template-args",
+          message: "Template argument 'U' is required and not specified.",
+          pos,
+          end,
+        },
+      ]);
+    });
+
+    it("cannot specify positional argument after named argument with default omitted", async () => {
+      const { pos, end, source } = extractSquiggles(`
+      model A<T, U = int32, V extends string = string> { a: T, b: U, c: V }
+
+      @test model B {
+        foo: A<boolean, V = "bar", ~~~string~~~>
+      }
+    `);
+
+      testHost.addTypeSpecFile("main.tsp", source);
+
+      const [{ B }, diagnostics] = (await testHost.compileAndDiagnose("main.tsp")) as [
+        { B: Model },
+        Diagnostic[],
+      ];
+
+      const foo = B.properties.get("foo")!.type;
+      strictEqual(foo.kind, "Model");
+      const a = foo.properties.get("a")!;
+      const b = foo.properties.get("b")!;
+      const c = foo.properties.get("c")!;
+      strictEqual(a.type.kind, "Scalar");
+      strictEqual(a.type.name, "boolean");
+      strictEqual(b.type.kind, "Scalar");
+      strictEqual(b.type.name, "int32");
+      strictEqual(c.type.kind, "String");
+      strictEqual(c.type.value, "bar");
+
+      expectDiagnostics(diagnostics, {
+        code: "invalid-template-args",
+        message:
+          "Positional template arguments cannot follow named arguments in the same argument list.",
+        pos: pos,
+        end: end,
+      });
+    });
+
+    it("decorators are applied in the correct order", async () => {
+      const members: [Type, Type][] = [];
+
+      testHost.addJsFile("effect.js", {
+        $effect: (_: DecoratorContext, target: Type, value: Type) => {
+          members.push([target, value]);
+        },
+      });
+
+      testHost.addTypeSpecFile(
+        "main.tsp",
+        `
+        import "./effect.js";
+
+        @effect(T)
+        model Dec<T> { t: T }
+
+        model A<T, U> { a: T, b: U }
+
+        @test model B {
+          bar: A<U = Dec<string>, T = Dec<int32>>
+        }
+        `
+      );
+
+      const { B } = (await testHost.compile("main.tsp")) as { B: Model };
+      const bar = B.properties.get("bar")!.type;
+      strictEqual(bar.kind, "Model");
+      const a = bar.properties.get("a")!;
+      const b = bar.properties.get("b")!;
+      strictEqual(a.type.kind, "Model");
+      strictEqual(a.type.name, "Dec");
+      strictEqual(b.type.kind, "Model");
+      strictEqual(b.type.name, "Dec");
+
+      // Assert that the members are added (decorators executed) in _declaration_ order
+      // rather than in the order they appear in the template instantiation.
+      strictEqual(members.length, 2);
+      strictEqual(members[0][0], a.type);
+      strictEqual(members[0][1].kind, "Scalar");
+      strictEqual(members[0][1].name, "int32");
+      strictEqual(members[1][0], b.type);
+      strictEqual(members[1][1].kind, "Scalar");
+      strictEqual(members[1][1].name, "string");
+    });
+  });
```

Comment:
Can you add a test if you specify the same parameter via positional and then named

Reply:
And also a test to make sure we have no issue with `model Foo<A, B = A>` then you instantiate A with named `Foo<A = string>`

Reply:
> Can you add a test if you specify the same parameter via positional and then named

I was actually writing this test while you were reviewing. I'll add the other as well.